Mathlib/Algebra/Module/Submodule/Pointwise.lean

```diff
@@ -277,4 +289,221 @@ scoped[Pointwise] attribute [instance] Submodule.pointwiseMulActionWithZero
+    (closed_under_smul : ∀ ⦃r : R⦄ ⦃n : M⦄, r ∈ s → n ∈ N → r • n ∈ p) :
+    s • N ≤ p :=
+  sInf_le closed_under_smul
+
+lemma set_smul_submodule_eq_of_le (p : Submodule R M)
+    (closed_under_smul : ∀ ⦃r : R⦄ ⦃n : M⦄, r ∈ s → n ∈ N → r • n ∈ p)
+    (le : p ≤ s • N) :
+    s • N = p :=
+  le_antisymm (set_smul_submodule_le s N p closed_under_smul) le
+
+lemma set_smul_submodule_mono_right {p q : Submodule R M} (le : p ≤ q) :
+    s • p ≤ s • q :=
+  set_smul_submodule_le _ _ _ fun _ _ hr hm => mem_set_smul_submodule_of_mem_mem (mem1 := hr)
+    (mem2 := le hm)
+
+lemma set_smul_submodule_mono_left {s t : Set R} (le : s ≤ t) :
+    s • N ≤ t • N :=
+  set_smul_submodule_le _ _ _ fun _ _ hr hm => mem_set_smul_submodule_of_mem_mem (mem1 := le hr)
+    (mem2 := hm)
+
+lemma set_smul_submodule_inductionOn {prop : M → Prop} (x : M)
+    (hx : x ∈ s • N)
+    (smul₀ : ∀ ⦃r : R⦄ ⦃n : M⦄, r ∈ s → n ∈ N → prop (r • n))
+    (smul₁ : ∀ (r : R) ⦃m : M⦄, prop m → prop (r • m))
+    (add : ∀ ⦃m₁ m₂ : M⦄, prop m₁ → prop m₂ → prop (m₁ + m₂))
+    (zero : prop 0) :
+    prop x :=
+  set_smul_submodule_le s N {
+    carrier := {m : M | prop m}
+    add_mem' := λ {x y} ↦ @add x y
+    zero_mem' := zero
+    smul_mem' := smul₁
+  } smul₀ hx
+
+lemma set_smul_submodule_eq_submodule_map [SMulCommClass R R N] :
+    s • N =
+    Submodule.map
+      (N.subtype.comp (Finsupp.lsum R λ (r : R) ↦ ⟨⟨(r • .), smul_add r⟩,
+        λ (r' : R) (m : N) ↦ by dsimp; rw [smul_comm]⟩))
+      (Finsupp.supported N R s) := by
+  classical
+  apply set_smul_submodule_eq_of_le
+  · intro r n hr hn
+    exact ⟨Finsupp.single r ⟨n, hn⟩, Finsupp.single_mem_supported _ _ hr, by simp⟩
+
+  · intro x hx
+    obtain ⟨c, hc, rfl⟩ := hx
+    simp only [LinearMap.coe_comp, coeSubtype, Finsupp.coe_lsum, Finsupp.sum, LinearMap.coe_mk,
+      AddHom.coe_mk, Function.comp_apply, AddSubmonoid.coe_finset_sum, coe_toAddSubmonoid,
+      SetLike.val_smul]
+    refine Submodule.sum_mem (p := s • N) (t := c.support) ?_ _ ⟨s • N, ?_⟩
+    · rintro r hr
+      rw [mem_set_smul_submodule_def, Submodule.mem_sInf]
+      rintro p hp
+      exact hp (hc hr) (c r).2
+    · ext x : 1
+      simp only [Set.mem_iInter, SetLike.mem_coe]
+      fconstructor
+      · refine λ h ↦ h λ r n hr hn ↦ ?_
+        rw [mem_set_smul_submodule_def, mem_sInf]
+        exact λ p hp ↦ hp hr hn
+      · exact λ h _ ↦ h
+
+lemma mem_set_smul_submodule (x : M) [SMulCommClass R R N] :
+    x ∈ s • N ↔ ∃ (c : R →₀ N), (c.support : Set R) ⊆ s ∧ x = c.sum λ r m ↦ r • m := by
+  fconstructor
+  · intros h
+    rw [set_smul_submodule_eq_submodule_map] at h
+    obtain ⟨c, hc, rfl⟩ := h
+    exact ⟨c, hc, rfl⟩
+
+  · rw [mem_set_smul_submodule_def, Submodule.mem_sInf]
+    rintro ⟨c, hc1, rfl⟩ p hp
+    simp only [Finsupp.sum, AddSubmonoid.coe_finset_sum, coe_toAddSubmonoid, SetLike.val_smul]
+    exact Submodule.sum_mem _ λ r hr ↦ hp (hc1 hr) (c _).2
+
+@[simp] lemma mem_empty_smul_submodule (x : M) : x ∈ (∅ : Set R) • N ↔ x = 0 := by
```

Comment:
I think a better simp lemma would be `(∅ : Set R) • N = ⊥` (or both if there would be DTT issues)

Reply:
changed to `empty_smul_submodule`